src/librustc_resolve/lib.rs

```diff
@@ -4200,7 +4200,7 @@ impl<'a, 'crateloader: 'a> Resolver<'a, 'crateloader> {
                         if self.session.rust_2018() && !in_module_is_extern {
                             // crate-local absolute paths start with `crate::` in edition 2018
                             // FIXME: may also be stabilized for Rust 2015 (Issues #45477, #44660)
-                            if graph_root {
+                            if name == keywords::Crate.name() {
                                 segms.insert(
                                     0, ast::PathSegment::from_ident(keywords::Crate.ident())
                                 );
```

Comment:
I don't think we want this `if name == keywords::Crate.name()`, but what exactly we *do* want is a bit tricky. If we were just coding for the Rust 2018 stuff, I think we would could simplify to

```rust
segms.insert(0, ast::PathSegment::from_ident(name));
```

full stop. The problem is that, then, if we are walking in an extern crate like `extern crate my_crate`, we're going to get suggestions like

```
crate::my_crate::bar
```

Actually, this suggestion is **totally correct**, but maybe we want to nudge people towards the shorter `my_crate::...` if we can.

Looking more closely at this code, actually, I think the current setup will also do the wrong thing if the `extern crate` is not at the root? We should make a test about *that*. e.g.

```rust
mod foo {
    extern crate my_crate;
}
```

In Rust 2018, I would expect to see a suggestion either like `my_crate::...` or `crate::foo::my_crate::...`. It looks to me like I will get `foo::my_crate::...`, because of the `!in_module_is_extern` check above.

Anyway, having thought it over some, I think the logic should probably be just the simple thing:

```rust
if self.session.rust_2018() {
    segms.insert(0, ast::PathSegment::from_ident(name));
}
```

The main downside to this is that, if you have an `extern crate`, I think you will get two suggestions: 

```
my_crate::...
crate::my_crate::...
```

both of which are correct, of course. To stop *that*, I think the easiest thing would be to just not walk into an `extern crate` in Rust 2018 mode, since we're going to walk the extern crates later when we walk the prelude crates...? Or, at least, don't walk extern crates that are also in the prelude.

Reply:
To stop that, we need to modify the code below:

```rust
if !in_module_is_extern || name_binding.vis == ty::Visibility::Public {
                         // add the module to the lookup
                         let is_extern = in_module_is_extern || name_binding.is_extern_crate();
                         if seen_modules.insert(module.def_id().unwrap()) {
                             worklist.push((module, path_segments, is_extern));
                         }
}
```

to include a check like

```rust
let is_extern_crate_that_also_appears_in_prelude = 
  name_binding.is_extern_crate() &&
  self.extern_prelude.contains(&ident);

let is_visible_to_user = 
  !in_module_is_extern || name_binding.vis == ty::Visibility::Public;

if 
    i!s_extern_crate_that_also_appears_in_prelude
    && is_visible_to_user
{
    let is_extern = in_module_is_extern || name_binding.is_extern_crate();
    if seen_modules.insert(module.def_id().unwrap()) {
        worklist.push((module, path_segments, is_extern));
    }
}
```